ls1intum/Artemis
# `Development`: Add script to generate a code coverage table automatically

### Checklist
#### General



### Motivation and Context



Filling out the code coverage table of PRs has been very time consuming and cumbersome, especially for larger pull requests. Some automation in that regard would save lots of time and ensure that the code coverage table is complete and correct.

@Mtze requested this script to be added to the Artemis repository in the `supporting_scripts` top level directory, which as it turns out does not exist.

### Description


Added a `supporting_scripts` top level directory with a `generate_code_cov_table` subdirectory containing the python script to automatically generate a code coverage table from the Bamboo build artefacts.

The script requires you to provide Bamboo login credentials and by default compares the changed files between a selected branch head (default: checked out branch) and its base commit (branching point from `origin/develop`). Afterwards, the Bamboo build plan artefacts (default: latest build number) are crawled for each changed file and the line coverage is extracted. Finally, a table according to the PR template is created and stored in the clipboard.

Additionally, added a comment in the `PULL_REQUEST_TEMPLATE.md` hinting at the existence of this script.

### Steps for Testing

1. Follow the setup steps in the `supporting_scripts/generate_code_cov_table/README.md`
2. Example command: `python generate_code_cov_table.py --branch-name feature/programming-exercise/exercise-options-consolidation`
3. Try out different branches (both `client_tests` and `server_tests` have to pass for that branch, use `--info` for more information on how to change the Bamboo build number etc.)
4. Enable `--verbose` flag and check if skipped files and included files are correct
5. Past the table into a markdown viewer and check if it is correct

### Review Progress






#### Code Review

#### Manual Tests

### Screenshots



**Output of `python generate_code_cov_table.py --branch-name origin/chore/remove-server-api-url`:**
<img width="589" alt="image" src="https://github.com/ls1intum/Artemis/assets/5898705/6f2f9e6b-9f75-462a-99b1-1ab6f86ba04f">


**Example of a generated table:**
<img width="631" alt="image" src="https://user-images.githubusercontent.com/5898705/235191511-c3cbbe87-43ac-430b-bb3d-775b827401f3.png">
